specification/dartLangSpec.tex

```diff
@@ -16231,238 +16306,447 @@ \subsection{Imports}
-It is important that errors are given at the offending declaration so that the party that receives the error can respond to it a meaningful way.
+\rationale{%
+The purpose of having members of the imported library in
+\NamespaceName{\metavar{deferred}}
+is to ensure that usages of members that have not yet been loaded
+can be resolved normally and has a well-defined behavior,
+but will raise errors.%
+}
+
+\LMHash{}%
+After an invocation of \code{$p$.loadLibrary()} succeeds,
+the name $p$ is mapped to a non-deferred prefix run-time namespace
+\NamespaceName{\metavar{loaded}},
+with bindings as described below for immediate imports.
+In addition, \NamespaceName{\metavar{loaded}}
+maps \code{loadLibrary} to a function with the same signature as before,
+and so it is possible to invoke \code{$p$.loadLibrary()} again.
+If a call fails, the library has not been loaded,
+and one has the option to invoke \code{$p$.loadLibrary()} again.
+Whether a repeated call to \code{$p$.loadLibrary()} succeeds will vary,
+as described below.
 
-In contrast a library comprises both imports and their usage; the library is under the control of a single party and so any problem stemming from the import can be resolved even if it is reported at the use site.%
+\commentary{%
+Note that it is a compile-time error for a deferred prefix
+to be used in more than one import,
+which means that the update of the binding of $p$
+does not interfere with other imports.%
 }
 
-\commentary{
-Note that no errors are raised if one hides or shows a name that is not in a namespace.
-}
-\rationale{
-This prevents situations where removing a name from a library would cause breakage of a client library.
+\LMHash{}%
+The effect of a repeated invocation of \code{$p$.loadLibrary()} is as follows:
+\begin{itemize}
+\item
+  If another invocation of \code{$p$.loadLibrary()} has already succeeded,
+  the repeated invocation also succeeds.
+  Otherwise,
+\item
+  If another invocation of \code{$p$.loadLibrary()} has failed:
+  \begin{itemize}
+  \item
+    If the failure is due to a compilation error,
+    the repeated invocation fails for the same reason.
+  \item
+    If the failure is due to other causes,
+    the repeated invocation behaves as if no previous call had been made.
+  \end{itemize}
+\end{itemize}
+
+\commentary{%
+In other words, one can retry a deferred load after a network failure
+or because a file is absent,
+but once one finds some content and loads it, one can no longer reload.
+We do not specify which object the returned future resolves to.%
 }
+\EndCase
 
 \LMHash{}%
-The dart core library \code{dart:core} is implicitly imported into every dart library other than itself via an import clause of the form
+\Case{Semantics of immediate imports}
+When $I_i$ is an immediate import that refers to a URI via the string $s_i$:
+Let $\ell_i$ denote the run-time entity that corresponds to the result of
+compiling the library $L_i$ denoted by $s_i$.
```

Comment:
Or maybe, as an overall statement:

> In Dart source code, libraries are denoted and identified by URIs. The URI is both used as identifier for the library and to locate the source code of the library. (Library parts are also denoted by URIs, but those URIs are only used to locate the part's source code). 
> The same URI used in different imports or exports denote the *same* library. Different URIs always denote *distinct* libraries. That means that a library uniquely determines the URI that denotes it. We call that "the URI of the library".
> _Since a library can have mutable state, it matters whether two imports denote the same library or different libraries with._.
> An import or export statement exposes declarations from the denoted library's export scope. 
> A deferred import of a library *L* with prefix *p* requires <code>*p*.loadLibrary()<code> to be successfully invoked (the returned future completes without an error) before the declarations exported by *L* can be safely accessed through *p*.  It may be a run-time error to access a declaration through *p* until the `loadLibrary` invocation is successful.

Reply:
Ah, the link in the "review arrived" email took me to a view showing comments on a commit from Aug 18 (made today). With a reload showing differences from all commits I suddenly get more comments (including this one). Responding to those now.

Reply:
> What is the difference between ℓi and Li?

Good question! I think we should aim at an improvement in the "natural language type correctness" here, but we also need to recognize that we won't specify anything to the bottom, there will be a number of layers in the real running program/hardware that we do not attempt to say anything about.

I think the main distinction is that you are thinking about source code and compilation; I'm thinking about a library as an entity that works similarly to the program representation in a compiler front end. That part is needed in order to be able to specify the static analysis. Then I'm including one more step where the front end output is compiled to an executable form; you seem to think that this is not needed, but I think we can't talk about an `object` without mentioning (without details) that programs can be executed, and the compiled representation of a function is not the same thing as the representation of the function that we talk about during static analysis.

The front end representation of a library is what we usually call _L_. The $\ell$ is the representation of the library which is needed to enable execution.

> So, let's define our domain entities:
> 
> - URIs, which point to
> - Source files, which are compiled into
> - Libraries.

I don't see a need to say much about the source code. We specify the grammar, which implies a certain structure on ASTs, and then we specify the static analysis and semantics in terms of ASTs.

So I agree that we have those entities (URI, source, library), but I think we need to recognize that `object` fits together with `the result of compiling a function` and similar run-time entities, and if we refuse to make a distinction between entities used to specify the static analysis and entities participating in an execution then we will actually make it harder to say that there _is_ a notion of 'compile-time' which is not run time.

I've edited the text to include the uniqueness rules (URI <=> library). However, wouldn't we need to take an implementation dependent normalization step into account, such that implementations are free to reduce `foo/../bar` to `./bar` and such?

I've also edited the text to maintain the distinction between the different types of entities more explicitly. But I've maintained the distinction between compile time and run time, and I'm still saying nothing about the nature of "the result of compiling a function" and so on. I do think that's a reasonable place to stop.

Reply:
The way I see it, our job is to specify the semantics of source code, preferably without restraining implementations unnecessarily.
Anything intermediate, including ASTs, is a tool we have given to ourselves, an abstraction layer to help ourselves, but not something inherent to the language. Someone else could specify the same thing in a completely different way (well, they could try!), and still end specifying the same language.

The "result of compiling a function" is not something we *need* to have in our model. The word that stands out to me (like a sore thumb) is "compiling". That's an implementation detail. We should specify that a function declaration evaluates to an object, and which properties that object has, but not that there is necessarily any compiling done along the way.

For URIs, yes we need to say something like: mumble mumble equivalence mumble URIs mumble. 
*Equivalent* URIs designate the *same* library. Equivalence is (currently) based on normalization. I'd like to allow frontends to have wider equivalences too where it makes sense (basically, if it's recognizably the same file, you should only need to read it once). For now, let me read what we have and see where we can squeeze an "equivalent" into it.

Reply:
> our job is to specify the semantics of source code .. [and nothing intermediate]

I agree that any intermediate layers between the source code and the observable behavior of a program execution can easily be an overspecification. But I do not think it would be helpful (or that it would yield a readable specification) if, for instance, we were to insist that we cannot mention objects, their run-time type, functions (both function objects and the notion that there is _something_ at run-time which can be executed), and so on.

If you agree that we can mention objects and so on, the remaining issue is mainly the word 'compiling'. I changed the wording to be more generic, thus allowing for compilers and interpreters and other technologies, including an interpreter which will actually work directly on the source code. I think this wording doesn't imply any specific decisions about implementation details.